### PR TITLE
Run nightly builds in dagster-world

### LINKED ISSUE #2296: Convert nightly build scripts from legacy to dagster
The CLI commands which are used in the nightly build scripts have changed a little bit so we need to update the scripts. Changes:

- The clobber arguments no longer exist for all of the commands because dagster just replaces the tables instead of deleting the whole database.
- No more bypass local cache?
- pudl_etl not runs the cems pipeline so no need to run epacems_to_parquet

Scope:
- [ ] local_pudl_etl.sh works in Docker
- [ ] gcp_pudl_etl.sh publishes valid outputs in intake.catalyst-coop

Next steps:
- [x] run local_pudl_etl.sh in a docker instance on a GCE VM

### PR DESCRIPTION
**Pull request by jdangerx**
Closes #2296 

Still waiting on full production run to succeed, but it worked with a subset of the data.

Main changes:
* add `PUDL_OUTPUT`, `PUDL_CACHE`, `DAGSTER_HOME` env vars to scripts
* make `DAGSTER_HOME` directory
* slightly streamline docker build process (mamba env builds first - this allows for more build caching, docker image builds on the developer box went from 4m to 2m)
* `set -eux` in the scripts

As you can see from the commit history, lots of typos etc. because it was hard to test this locally :(

Here are some thoughts for the future:

- [ ] consolidate local/gcp scripts - what works in one may not work in the other. Lots of typos occurred in porting changes from one place to another. If we do Docker development on a VM in GCP, all the google auth stuff should work - we can just have a flag that says "don't send things to intake in GCS or S3," but have everything else be the same
- [ ] build a `requirements.txt` so that we can install all dependencies earlier in the Docker build, allowing for faster caching
- [ ] add a flag to "go really frickin fast and skip a lot of data" so testing is faster
- [x] document the "log into vm, find docker image id via `docker ps` or `journalctl`, `docker exec -it {id} bash and poke around post-failure`" workflow somewhere
- [ ] run epacems_to_parquet and ferc_to_sqlite/pudl_etl in parallel

### CODE REVIEW

File: .github/workflows/build-deploy-pudl.yml [hunk 1]
```diff
@@ -70,6 +70,8 @@ jobs:
           push: ${{ github.event_name != 'pull_request' }}
           tags: ${{ steps.docker_metadata.outputs.tags }}
           labels: ${{ steps.docker_metadata.outputs.labels }}
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
```
Comment (bendnorman) [inline .github/workflows/build-deploy-pudl.yml:74]: What is this doing?

File: docker/Dockerfile [hunk 2]
```diff
@@ -33,24 +31,29 @@ ENV CONDA_RUN="conda run --no-capture-output --prefix ${CONDA_PREFIX}"
...
 # Create a conda environment based on the specification in the repo
+COPY test/test-environment.yml test/test-environment.yml
+RUN mamba create --copy --prefix ${CONDA_PREFIX} --yes python=${PYTHON_VERSION} && \
+    # Then we can use mamba env update, which can parse the environment.yml file:
+    mamba env update --prefix ${CONDA_PREFIX} --file test/test-environment.yml && \
+    conda clean -afy
+
+# Copy the cloned pudl repository into the user's home directory
+COPY --chown=catalyst:catalyst . ${CONTAINER_HOME}
+
```
Comment (bendnorman) [inline docker/Dockerfile:45]: Why did you move this? I assume it enables more image layer caching?

File: src/pudl/convert/censusdp1tract_to_sqlite.py [hunk 3]
```diff
@@ -62,7 +62,7 @@ def censusdp1tract_to_sqlite(pudl_settings=None, year=2010, ds=None, clobber=Fal
         tmpdir_path = Path(tmpdir)
         zip_ref = ds.get_zipfile_resource("censusdp1tract", year=year)
         extract_root = tmpdir_path / Path(zip_ref.filelist[0].filename)
-        out_path = Path(pudl_settings["sqlite_dir"]) / "censusdp1tract.sqlite"
+        out_path = Path(pudl_settings["pudl_out"]) / "censusdp1tract.sqlite"
```
Comment (bendnorman) [inline src/pudl/convert/censusdp1tract_to_sqlite.py:65]: Ah good catch!

**Review by bendnorman — APPROVED** [review]
Looks good to me! I mostly left questions for you. 

I like your future work ideas.
- Consolidating scripts: I agree it's not ideal having the two scripts. I've found we haven't really been using the docker stuff locally. Would probably be better to design these scripts so most of the testing happens in a GCP VM.
- requirements.txt file: Do you think we should only use a requirements file without a mamba env? It'd probably be fast but it is nice testing our the conda env.
- fast data: can use just use the fast ETL?
- CEMS: We can't run ferc_to_sqlite, pudl_etl, and epacems in parallel because there are dependencies between all of them. I think the solution to speed this up is to parallelize the CEMS processing. Once dagster-asset-etl is in dev and we're rolling on the output table conversions I was going to experiment more with using [dagster partitions](https://docs.dagster.io/concepts/partitions-schedules-sensors/partitions) to parallelize CEMS.

File: .github/workflows/build-deploy-pudl.yml (hunk 1 above)
Comment (jdangerx) [reply .github/workflows/build-deploy-pudl.yml:74]: This is caching docker build layers in the GH actions cache, since otherwise the runners have to rebuild everything from scratch every time.

File: docker/Dockerfile (hunk 2 above)
Comment (jdangerx) [reply docker/Dockerfile:45]: Yep! Basically, since `.` changes all the time, pulling in `.` is going to invalidate the cache for everything after it. But we can copy in the requirements files first, and those don't change as frequently. This lets us skip the ~2m mamba build step every time I try to rebuild the image.

File: src/pudl/convert/censusdp1tract_to_sqlite.py (hunk 3 above)
Comment (jdangerx) [reply src/pudl/convert/censusdp1tract_to_sqlite.py:65]: I think there might be a few other spots where we're confused about where to look for files... I'll wait until the prod tests fail to try to find them.

**Comment by bendnorman** [comment]
I just went through the logs and a couple of things don't look right. CEMS took 2 hrs and 40 minutes to run which is about twice as long as it use to take. Also we should be getting the DEBUG level logs.

File: src/pudl/workspace/setup.py [hunk 4]
```diff
@@ -229,7 +227,7 @@ def init(pudl_in, pudl_out, clobber=False):
     deploy(settings_pkg, settings_dir, ignore_files, clobber=clobber)
 
     # Make output directory:
-    pudl_out = pathlib.Path(pudl_settings["pudl_out"]) / "output"
+    pudl_out = pathlib.Path(pudl_settings["pudl_out"])
```
Comment (jdangerx) [inline src/pudl/workspace/setup.py:230]: @bendnorman I took out the `/ "output"` because it seemed like we were writing files directly to `$PUDL_OUTPUT/` and the test were looking for files in `$PUDL_OUTPUT/output` - but maybe we should be telling the ETL to write things to `$PUDL_OUTPUT` instead?

Comment (jdangerx) [reply src/pudl/workspace/setup.py:230]: FWIW this *is* uploading to intake correctly 🤷
